src/functions/env.ts

```diff
@@ -0,0 +1,203 @@
+import * as clc from "cli-color";
+
+import { FirebaseError } from "../error";
+import * as envstore from "./envstore";
+
+// Exported for testing-only
+export const ENVSTORE_ID = "firebase-functions";
+const RESERVED_KEY_NAMES = [
+  // Cloud Functions for Firebase
+  "FIREBASE_CONFIG",
+  // Cloud Functions:
+  //   https://cloud.google.com/functions/docs/env-var#best_practices_and_reserved_environment_variables
+  "FUNCTION_TARGET",
+  "FUNCTION_SIGNATURE_TYPE",
+  "K_SERVICE",
+  "K_REVISION",
+  "PORT",
+  // Cloud Run:
+  //   https://cloud.google.com/run/docs/reference/container-contract#env-vars
+  "K_CONFIGURATION",
+];
+
+/**
+ * Format environment variables into console-friendly strings.
+ *
+ * @param {Record<string, string>} envs Environment variables to format.
+ * @return {string} Formatted string suitable for printing.
+ */
+export function formatEnv(envs: Record<string, string>): string {
+  const s = [];
+  for (const [k, v] of Object.entries(envs)) {
+    s.push(`${k}=${v}`);
+  }
+  return s.join("\n");
+}
+
+/**
+ * Validates string for use as an env var key.
+ *
+ * We restrict key names to ones that conform to POSIX standards.
+ * This is more restrictive than what is allowed in Cloud Functions or Cloud Run.
+ *
+ * @param key {string} Key to validate
+ */
+export function validateKey(key: string): void {
+  // key must not be one of the reserved key names.
+  if (RESERVED_KEY_NAMES.includes(key)) {
+    throw new FirebaseError(
+      "Invalid environment variable name " + clc.bold(key) + ", is reserved for internal use."
+    );
+  }
+  // Only allow subset of key names that conforms to POSIX standards for
+  // environment variables:
+  //   https://cloud.google.com/functions/docs/env-var#portability
+  if (!/^[A-Z_][A-Z0-9_]+$/.test(key)) {
+    throw new FirebaseError(
+      "Invalid environment variable name " +
+        clc.bold(key) +
+        ", must start with an uppercase ASCII letter or underscore" +
+        ", and then consist of uppercase ASCII letters, digits, and underscores."
+    );
+  }
+  // Keys cannot contain the prefix X_GOOGLE_.
+  if (/^X_GOOGLE_.*$/.test(key)) {
+    throw new FirebaseError(
+      "Invalid environment variable name " +
+        clc.bold(key) +
+        ", cannot contain the prefix X_GOOGLE_."
+    );
+  }
+}
+
+/**
+ * Parse list of strings in key=value format as an object.
+ *
+ * @param {string[]} args List of strings in key=value format.
+ * @return {Record<string, string>} An object with key, value pairs.
+ */
+export function parseKvArgs(args: string[]): Record<string, string> {
+  const envs: Record<string, string> = {};
+  args.forEach((arg) => {
+    const parts = arg.split("=");
+    if (parts.length < 2) {
+      throw new FirebaseError(`Invalid argument ${clc.bold(arg)}, must be in key=val format`);
+    }
+    const key = parts[0];
+    validateKey(key);
+    const val = parts.slice(1).join("="); // Val may have contained '='.
```

Comment:
You can avoid having to do this by adding a limit to the split: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split

Reply:
Hmm this didn't work as expected:

```js
const w = "ABC=EF=G"
console.log(w.split("="))
=> ["ABC", "EF", "G"] 
console.log(w.split("=", 2))
=> ["ABC", "EF"] 
```
[link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBA7jAvDARAQQEIGFEFEBiiA4igFCiQgA2ApgHRUgDmAFHHRAA5UCWULKRCgCUw8uAjV6jVuy69+glABoYAJlFA)

From docs:

> limit Optional
A non-negative integer specifying a limit on the number of substrings to be included in the array. If provided, splits the string at each occurrence of the specified separator, but stops when limit entries have been placed in the array. **Any leftover text is not included in the array at all**.

Am I holding this correctly?

Reply:
Ahhh, i dint realize it tossed away the rest. Your original way seems best then!